spec/index.bs

```diff
@@ -569,6 +561,63 @@ enum GPUBindingType {
+
+<div algorithm="GPUDevice.createBindGroupLayout(descriptor)">
+Any time during this algorithm, to <dfn>return with error</dfn> is to:
+    1. Generate a validation error in the current scope with appropriate error message.
+    1. Create a new invalid {{GPUBindGroupLayout}} and return the result.
+
+The <dfn method for "GPUDevice">createBindGroupLayout(|descriptor|)</dfn> method attempts to create a valid {{GPUBindGroupLayout}}.
+
+1. If the device is [=lost=], [=return with error=].
+1. Let |layout| be a new valid {{GPUBindGroupLayout}} object.
+1. For each {{GPUBindGroupLayoutBinding}} |bindingDescriptor| in |descriptor|.{{GPUBindGroupLayoutDescriptor/bindings}}:
+    1. If |bindingDescriptor|.{{GPUBindGroupLayoutBinding/binding}} is already present in |layout|'s [=internal map=], [=return with error=].
+    1. Let |layoutBinding| be a new internal [=layout binding=] object.
+    1. If |bindingDescriptor|.{{GPUBindGroupLayoutBinding/type}} is {{GPUBindingType/uniform-buffer}}:
+        1. If |layout|'s [=internal map=] contains {{GPULimits/maxUniformBuffersPerShaderStage|GPULimits.maxUniformBuffersPerShaderStage}} [=layout binding=]s of [=type=] {{GPUBindingType/uniform-buffer}} that all share any [=visibility=] bit contained in |bindingDescriptor|.{{GPUBindGroupLayoutBinding/visibility}}, [=return with error=].
+        1. Set [=dynamic=] of |layoutBinding| to the value of |bindingDescriptor|.{{GPUBindGroupLayoutBinding/dynamic}}.
+    1. If [=type=] is {{GPUBindingType/storage-buffer}} or {{GPUBindingType/readonly-storage-buffer}}:
+        1. If |layout|'s [=internal map=] contains {{GPULimits/maxStorageBuffersPerBindGroup|GPULimits.maxStorageBuffersPerBindGroup}} [=layout binding=]s of [=type=] {{GPUBindingType/storage-buffer}} or {{GPUBindingType/readonly-storage-buffer}}, [=return with error=].
+        1. Set [=dynamic=] of |layoutBinding| to the value of |bindingDescriptor|.{{GPUBindGroupLayoutBinding/dynamic}}.
+    1. If [=type=] is {{GPUBindingType/sampled-texture}} or {{GPUBindingType/storage-texture}}:
+        1. If [=type=] is {{GPUBindingType/sampled-texture}} and if |layout|'s [=internal map=] contains {{GPULimits/maxSampledTexturesPerShaderStage|GPULimits.maxSampledTexturesPerShaderStage}} [=layout binding=]s of [=type=] {{GPUBindingType/sampled-texture}} that all share any [=visibility=] bit contained in |bindingDescriptor|.{{GPUBindGroupLayoutBinding/visibility}}, [=return with error=].
+        1. If [=type=] is {{GPUBindingType/storage-texture}} and if |layout|'s [=internal map=] contains {{GPULimits/maxStorageTexturesPerShaderStage|GPULimits.maxStorageTexturesPerShaderStage}} [=layout binding=]s of [=type=] {{GPUBindingType/storage-texture}} that all share any [=visibility=] bit contained in |bindingDescriptor|.{{GPUBindGroupLayoutBinding/visibility}}, [=return with error=].
```

Comment:
Seems like we should also say "if dynamic is set, return with error", and similar for the other cases too.
(Unfortunate, because it's kind of verbose; maybe it can be condensed.)

IDL wise, I wonder if there's any way we can split GPUBindGroupLayoutBinding into 4 different dictionaries.

Reply:
I'd also prefer that we split these GPUBindGroupLayoutBinding cases into different sub-classes/dictionaries. However, if we keep it as is, I propose we ignore "dynamic" if the resource is not a buffer (maybe issue a warning?), and so on, instead of aborting `GPUBindGroupLayout` creation. It keeps the API more flexible.